engine/src/exec-interface-image.cpp

```diff
@@ -431,45 +431,68 @@ void MCImage::SetText(MCExecContext& ctxt, MCDataRef p_text)
 
 void MCImage::GetImageData(MCExecContext& ctxt, MCDataRef& r_data)
 {
-	// IM-2013-02-07: image data must return a block of data, even if the image is empty.
+    if (m_rep && m_rep->GetType() != kMCImageRepMutable && m_rep->IsLocked())
+    {
+        ctxt . LegacyThrow(EE_IMAGE_MUTABLELOCK);
+        return;
+    }
+    
+    bool t_locked = false;
+    if (m_rep)
+    {
+        t_locked = m_rep->GetType() == kMCImageRepMutable && m_rep->IsLocked();
+    }
+    
+    // IM-2013-02-07: image data must return a block of data, even if the image is empty.
 	// image data should always be for an image the size of the rect.
-	uint32_t t_pixel_count = rect.width * rect.height;
-	uint32_t t_data_size = t_pixel_count * sizeof(uint32_t);
-	
-	MCAutoByteArray t_buffer;
-	
-	bool t_success = true;
-	
-	t_success = t_buffer.New(t_data_size);
-	
-	if (t_success)
-	{
-		uint32_t *t_data_ptr = (uint32_t*)t_buffer.Bytes();
-		if (m_rep == nil)
-			MCMemoryClear(t_data_ptr, t_data_size);
-		else
-		{
+    uint32_t t_pixel_count = rect.width * rect.height;
+    uint32_t t_data_size = t_pixel_count * sizeof(uint32_t);
+    
+    MCAutoByteArray t_buffer;
+    
+    bool t_success = true;
+    
+    t_success = t_buffer.New(t_data_size);
+    
+    if (t_success)
+    {
+        uint32_t *t_data_ptr = (uint32_t*)t_buffer.Bytes();
+        if (m_rep == nil)
+            MCMemoryClear(t_data_ptr, t_data_size);
+        else
+        {
             // SN-2014-01-31: [[ Bug 11462 ]] Opening an image to get its data should not
             // reset its size: F_LOCK_LOCATION ensures the size - and the location, which
             // doesn't matter here - are read as they are stored.
             bool t_tmp_locked;
             t_tmp_locked = false;
             
-            if (!getflag(F_LOCK_LOCATION))
+            if (!t_locked)
```

Comment:
I think we should be able to make this:
```
MCImageBitmap *t_bitmap = nullptr;
bool t_tmp_locked_location = false;
if (m_rep->GetType() == kMCImageRepMutable)
{
  t_success = static_cast<MCMutableImageRep *>(m_rep)->LockBitmap(0, 0, t_bitmap);
}
else
{
  -- make sure locklocation is set
  -- make sure image is open
  t_success = lockbitmap(t_bitmap, false);
}

... make image data ...

if (m_rep->GetType() == kMCImageRepMutable)
{
  UnlockBitmap
}
else
{
  unlockbitmap
  closeimage
  reset locklocation
}
```
The rationale here is that if the image has a mutableimagerep, then we don't need to open or do anything else, as we are using the bitmap direct from the mutableimagerep (which will always have one).